In [2]:
# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES
# TO THE CORRECT LOCATION (/kaggle/input) IN YOUR NOTEBOOK,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.
import os, sys
from tempfile import NamedTemporaryFile
from urllib.request import urlopen
from urllib.parse import unquote, urlparse
from urllib.error import HTTPError
from zipfile import ZipFile
import tarfile, shutil

CHUNK_SIZE = 40960
DATA_SOURCE_MAPPING = 'iso-country-codes-global:https%3A%2F%2Fstorage.googleapis.com%2Fkaggle-data-sets%2F403474%2F773844%2Fbundle%2Farchive.zip%3FX-Goog-Algorithm%3DGOOG4-RSA-SHA256%26X-Goog-Credential%3Dgcp-kaggle-com%2540kaggle-161607.iam.gserviceaccount.com%252F20240303%252Fauto%252Fstorage%252Fgoog4_request%26X-Goog-Date%3D20240303T172247Z%26X-Goog-Expires%3D259200%26X-Goog-SignedHeaders%3Dhost%26X-Goog-Signature%3D41dba32620fc4be90053586d465b076719706aab250d851d66436c7106f38c56bbb4ea06e454bcfd9c57064d56a8f3edd60c9378e0f3e6a4e7a0949a107e37983857adcc8a85ff84312a5a08620ecb17ae14580435e7e6a84cf41c8c76cceb7edf3233a5dc507915c460dd3919362d3855f53a22883b6097bd5e98886713be35fd4718e36269808237ca06480139670cbeec3a5bb5ca9adad9d9eaf2cefc58b003a92a6f35fb05d46388be06a6966e4628beea3f21b45b8289c94442eb72d6d25729c521f50bd0fede3bf36ac58fa9c5090850c8c86e49ff71ef38721e4ce84a80ebbe7d5fe635e1ba9117c3059939adfbb189bc88a080b234705969d7b74c1b,violence-against-women-and-girls:https%3A%2F%2Fstorage.googleapis.com%2Fkaggle-data-sets%2F648963%2F1149239%2Fbundle%2Farchive.zip%3FX-Goog-Algorithm%3DGOOG4-RSA-SHA256%26X-Goog-Credential%3Dgcp-kaggle-com%2540kaggle-161607.iam.gserviceaccount.com%252F20240303%252Fauto%252Fstorage%252Fgoog4_request%26X-Goog-Date%3D20240303T172247Z%26X-Goog-Expires%3D259200%26X-Goog-SignedHeaders%3Dhost%26X-Goog-Signature%3D492acd3ac9f22ea58eaac31d470087318fed1b9aad9cfdbb9d867de9870cb1d14eed905afcd76f2f93117a51a1c8d85f6c1b23a33a43bf5e13836a1f90e7362bc8ce76cc1ab0845e17da45a075e08098bd2e3f4cca150346e660bd3f3aef0d82d60f4c5d623a442d334d1348518b9803c08c479f16a18dae5893182d9903e2b1d9d6f52f02a898c2648d11a606a004cd8fe395e48dd22bda8b87570f851ff5f97e5c06499c7604d4039247531f156678248d05f60d31951f0650b7f6ac22832a15fc5fe50fbe727eb851fa9f6cb1c234b3ee67dd495c54352e43ed84a40a5015285b992b2e234f9a8b376dce52aed8ca441f3dd786516c360e12d0139e645224'

KAGGLE_INPUT_PATH='/kaggle/input'
KAGGLE_WORKING_PATH='/kaggle/working'
KAGGLE_SYMLINK='kaggle'

!umount /kaggle/input/ 2> /dev/null
shutil.rmtree('/kaggle/input', ignore_errors=True)
os.makedirs(KAGGLE_INPUT_PATH, 0o777, exist_ok=True)
os.makedirs(KAGGLE_WORKING_PATH, 0o777, exist_ok=True)

try:
  os.symlink(KAGGLE_INPUT_PATH, os.path.join("..", 'input'), target_is_directory=True)
except FileExistsError:
  pass
try:
  os.symlink(KAGGLE_WORKING_PATH, os.path.join("..", 'working'), target_is_directory=True)
except FileExistsError:
  pass

for data_source_mapping in DATA_SOURCE_MAPPING.split(','):
    directory, download_url_encoded = data_source_mapping.split(':')
    download_url = unquote(download_url_encoded)
    filename = urlparse(download_url).path
    destination_path = os.path.join(KAGGLE_INPUT_PATH, directory)
    try:
        with urlopen(download_url) as fileres, NamedTemporaryFile() as tfile:
            total_length = fileres.headers['content-length']
            print(f'Downloading {directory}, {total_length} bytes compressed')
            dl = 0
            data = fileres.read(CHUNK_SIZE)
            while len(data) > 0:
                dl += len(data)
                tfile.write(data)
                done = int(50 * dl / int(total_length))
                sys.stdout.write(f"\r[{'=' * done}{' ' * (50-done)}] {dl} bytes downloaded")
                sys.stdout.flush()
                data = fileres.read(CHUNK_SIZE)
            if filename.endswith('.zip'):
              with ZipFile(tfile) as zfile:
                zfile.extractall(destination_path)
            else:
              with tarfile.open(tfile.name) as tarfile:
                tarfile.extractall(destination_path)
            print(f'\nDownloaded and uncompressed: {directory}')
    except HTTPError as e:
        print(f'Failed to load (likely expired) {download_url} to path {destination_path}')
        continue
    except OSError as e:
        print(f'Failed to load {download_url} to path {destination_path}')
        continue

print('Data source import complete.')


Le chemin d'accŠs sp‚cifi‚ est introuvable.


OSError: [WinError 1314] Le client ne dispose pas d’un privilège nécessaire: '/kaggle/input' -> '..\\input'

<h1><center>Violence Against Women and Girls</center></h1>

<center><img src="https://equineteurope.org/wp-content/uploads/2019/11/stop-violence-against-women.jpg" width="800"></img></center>



# Content

- <a href='#1'>Introduction</a>  
- <a href='#2'>Analysis preparation</a>  
- <a href='#3'>What demographics are telling us</a>   
- <a href='#4'>What the question answered reveals</a>
- <a href='#5'>Demographics and question answered combined</a>
- <a href='#6'>Median violence percent per country</a>  
- <a href='#7'>Maxim violence percent per country</a>
- <a href='#8'>Summary of findings</a>

# <a id='1'>Introduction</a>

We will analyze the data on violence on women and girls. The data is collected from 70 countries and contains demographics questions and answers as well as percent of response to 5 violence specific questions.



# <a id='2'>Analysis preparation</a>

We input the data, we glimpse the data, we look to missing data and we look to number of categories for each type of categorical data.

## Data input

In [3]:
import pandas as pd
data_df = pd.read_csv("C:/Users/hajar/Downloads/violence_data.csv")

## Glimpse the data

In [4]:
data_df.head()

RecordID      Country Gender Demographics Question  \
0         1  Afghanistan      F        Marital status   
1         1  Afghanistan      F             Education   
2         1  Afghanistan      F             Education   
3         1  Afghanistan      F             Education   
4         1  Afghanistan      F        Marital status   

          Demographics Response                   Question Survey Year  Value  
0                 Never married  ... if she burns the food  01/01/2015    NaN  
1                        Higher  ... if she burns the food  01/01/2015   10.1  
2                     Secondary  ... if she burns the food  01/01/2015   13.7  
3                       Primary  ... if she burns the food  01/01/2015   13.8  
4  Widowed, divorced, separated  ... if she burns the food  01/01/2015   13.8

In [5]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12600 entries, 0 to 12599
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   RecordID               12600 non-null  int64  
 1   Country                12600 non-null  object 
 2   Gender                 12600 non-null  object 
 3   Demographics Question  12600 non-null  object 
 4   Demographics Response  12600 non-null  object 
 5   Question               12600 non-null  object 
 6   Survey Year            12600 non-null  object 
 7   Value                  11187 non-null  float64
dtypes: float64(1), int64(1), object(6)
memory usage: 787.6+ KB


In [6]:
columns = data_df.columns
for column in columns:
    print(f"{column} : {data_df[column].nunique()}")

RecordID : 420
Country : 70
Gender : 2
Demographics Question : 5
Demographics Response : 15
Question : 6
Survey Year : 18
Value : 757


# <a id='3'>What demographics are telling us</a>


Let's look first to what demographics are telling to us. There are 5 demographics questions:
* Age (15-24, 25-34, 35-49)
* Education (No Education, Primary, Secondary, Higher)
* Employement (Unemployed, Employed for cash, Employed for kind)
* Marital status (Maried or living together, Widowed, divorced, separated, Never Married)
* Residence (Rural, Urban)

We will aggregate the data on the demographic question, answer and calculate the minimm, maximum, median and average stats per each of this demographic.

In [7]:
demographics_df = data_df.groupby(["Demographics Question", "Demographics Response"])["Value"].agg(["median", "max", "min", "mean"]).reset_index()
demographics_df.columns = ["Question", "Response", "Median", "Max", "Min", "Mean"]
print("Violence % median, min, max, and mean per demographic group")
demographics_df.sort_values(["Question", "Median"])

Violence % median, min, max, and mean per demographic group


Question                      Response  Median   Max  Min       Mean
2              Age                         35-49   14.15  81.0  0.2  19.336412
1              Age                         25-34   14.45  81.5  0.1  19.703562
0              Age                         15-24   17.50  80.1  0.1  21.084169
3        Education                        Higher    4.20  74.6  0.0   8.898670
6        Education                     Secondary   13.05  76.7  0.2  17.378892
5        Education                       Primary   18.40  80.5  0.1  22.819093
4        Education                  No education   21.55  82.0  0.0  25.403125
9       Employment                    Unemployed   14.55  80.1  0.0  19.539710
7       Employment             Employed for cash   14.85  81.5  0.1  19.553804
8       Employment             Employed for kind   20.15  86.9  0.3  24.445541
10  Marital status    Married or living together   14.80  82.1  0.2  20.181003
11  Marital status                 Never married   14.90  72.1  0.0  18.704360
12  Marital status  Widowed, divorced, separated   17.15  75.9  0.0  20.683511
14       Residence                         Urban   11.05  74.4  0.1  15.895778
13       Residence                         Rural   19.00  82.1  0.2  23.294063

The data grouped on demographics shows the following:
* The most exposed age group is the 15-24, with a median of 17.50 and max of 80.1, while the absolute max is for 25-34 age group;
* Education level of girls and women is a very good predictor for violence against them: a higher education level will ensure a small median percent of 4.2%, while No education shows a 21.55% median and a 82% maximum violence percent.
* Employment factor also counts, from Unemployed to Employed for kind the median is varying from 14.55% to 20.15%. The women that work tend to be more exposed to violence.
* Marriage is slighly protecting if we look to the median, but married women are more exposed to higher level of violence. Widowed, divorced or separated have a smaller median than Married or living together group, as well as the max. But it might be also an age factor here.
* Women from Urban areas are less exposed to violence than women from Rural areas, both the median and the max values being smaller in Urban vs. Rural areas.

In [8]:
data_df

RecordID      Country Gender Demographics Question  \
0             1  Afghanistan      F        Marital status   
1             1  Afghanistan      F             Education   
2             1  Afghanistan      F             Education   
3             1  Afghanistan      F             Education   
4             1  Afghanistan      F        Marital status   
...         ...          ...    ...                   ...   
12595       210     Zimbabwe      M             Residence   
12596       280     Zimbabwe      M             Residence   
12597       280     Zimbabwe      M             Residence   
12598       350     Zimbabwe      M             Residence   
12599       350     Zimbabwe      M             Residence   

              Demographics Response                                 Question  \
0                     Never married                ... if she burns the food   
1                            Higher                ... if she burns the food   
2                         Secondary                ... if she burns the food   
3                           Primary                ... if she burns the food   
4      Widowed, divorced, separated                ... if she burns the food   
...                             ...                                      ...   
12595                         Urban  ... if she goes out without telling him   
12596                         Rural         ... if she neglects the children   
12597                         Urban         ... if she neglects the children   
12598                         Rural  ... if she refuses to have sex with him   
12599                         Urban  ... if she refuses to have sex with him   

      Survey Year  Value  
0      01/01/2015    NaN  
1      01/01/2015   10.1  
2      01/01/2015   13.7  
3      01/01/2015   13.8  
4      01/01/2015   13.8  
...           ...    ...  
12595  01/01/2015   11.8  
12596  01/01/2015   20.1  
12597  01/01/2015   15.0  
12598  01/01/2015    7.2  
12599  01/01/2015    3.7  

[12600 rows x 8 columns]

**That's a lot better, now lets query results in order to make an exploratory data analysis acording to Age and Education.**

# <a id='4'>What the question answered reveals</a>

In [9]:
question_df = data_df.groupby(["Question"])["Value"].agg(["median", "max", "min", "mean"]).reset_index()
question_df.columns = ["Question", "Median", "Max", "Min", "Mean"]
print("Violence % median, min, max, and mean per question asked")
question_df.sort_values(["Median"])

Violence % median, min, max, and mean per question asked


Question  Median   Max  Min       Mean
2                ... if she burns the food     6.4  56.7  0.0   9.203445
5  ... if she refuses to have sex with him     9.0  68.7  0.0  13.209613
1               ... if she argues with him    15.7  76.5  0.0  18.983652
3  ... if she goes out without telling him    16.4  77.0  0.0  20.046321
4         ... if she neglects the children    20.8  75.6  0.0  23.461249
0     ... for at least one specific reason    31.0  86.9  0.0  33.217152

Generally, the median and maximum are aligned. The reasons for violence, ranged by median values are:
* if she burns the food;  
* if she refuses to have sex with him;  
* if she argues with him;  
* if she goes out without telling him;  
* if she neglects the children.  

Non specified but at least one specific reason would have the maximum median value and the absolute maximum percent.

# <a id='5'>Demographics and question answered combined</a>

In [10]:
demoq_df = data_df.groupby(["Demographics Question", "Demographics Response", "Question"])["Value"].agg(["median", "max", "min", "mean"]).reset_index()
demoq_df.columns = ["Demographics Question", "Demographics Response", "Question", "Median", "Max", "Min", "Mean"]
print("Violence % median, min, max, and mean per demographic group and question asked")
demoq_df = demoq_df.sort_values(["Demographics Question", "Demographics Response", "Median"])
demoq_df.head()

Violence % median, min, max, and mean per demographic group and question asked


Demographics Question Demographics Response  \
2                   Age                 15-24   
5                   Age                 15-24   
3                   Age                 15-24   
1                   Age                 15-24   
4                   Age                 15-24   

                                  Question  Median   Max  Min       Mean  
2                ... if she burns the food    8.50  47.7  0.1  10.098387  
5  ... if she refuses to have sex with him   10.05  59.4  0.3  13.404762  
3  ... if she goes out without telling him   18.40  67.7  0.2  21.026563  
1               ... if she argues with him   18.65  65.5  0.5  20.304032  
4         ... if she neglects the children   23.20  61.4  1.9  25.306250

In [11]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

def show_answers_per_demographic_question(question, aggregate="Median", layout_mode="Single"):
    font_size = 8
    title_font_size = 12
    if layout_mode == "Single":
        fig = go.Figure()
        font_size = 10
    else:
        fig = make_subplots(rows=2, cols=2, specs=[[{'type': 'polar'}]*2]*2)

    sel_df = demoq_df.loc[demoq_df["Demographics Question"]==question]
    answers = list(sel_df["Demographics Response"].unique())
    count = 0
    for answer in answers:
        subsel_df = sel_df.loc[sel_df["Demographics Response"] == answer]
        if layout_mode == "Single":
            fig.add_trace(go.Scatterpolar(name=answer, r=subsel_df[aggregate], theta=subsel_df['Question']))
        else:
            fig.add_trace(go.Scatterpolar(name=answer, r=subsel_df[aggregate], theta=subsel_df['Question']), count%2 + 1, int(count/2+1))
        count = count + 1
    fig.update_traces(fill='toself')
    fig.update_polars(bgcolor='white')
    fig.update_polars(angularaxis_linecolor='darkgrey')
    fig.update_polars(angularaxis_gridcolor='darkgrey')
    fig.update_polars(radialaxis_gridcolor='darkgrey')
    fig.update_polars(radialaxis_linecolor='darkgrey')
    fig.update_layout(title='Demografics Question: '+ question + ' | Statistics: ' + aggregate,font=dict(family="Courier New, monospace",size=title_font_size))
    fig.show()

In [12]:
show_answers_per_demographic_question("Age")

In [13]:
show_answers_per_demographic_question("Age", aggregate="Max")

In [14]:
show_answers_per_demographic_question("Education")

In [15]:
show_answers_per_demographic_question("Education", aggregate="Max")

In [16]:
show_answers_per_demographic_question("Employment")

In [17]:
show_answers_per_demographic_question("Employment", aggregate="Max")

In [18]:
show_answers_per_demographic_question("Marital status")

In [19]:
show_answers_per_demographic_question("Marital status", aggregate="Max")

In [20]:
show_answers_per_demographic_question("Residence")

In [21]:
show_answers_per_demographic_question("Residence", aggregate="Max")

# <a id='6'>Median violence percent per country</a>

In [29]:
country_df = pd.read_csv("C:/Users/hajar/Desktop/wikipedia-iso-country-codes.csv")
country_df.columns = ["Country", "A2C", "A3C", "Num","ISO"]
datac_df = data_df.merge(country_df)


In [30]:
country_df.head()

Country A2C  A3C  Num            ISO
0     Afghanistan  AF  AFG    4  ISO 3166-2:AF
1   Åland Islands  AX  ALA  248  ISO 3166-2:AX
2         Albania  AL  ALB    8  ISO 3166-2:AL
3         Algeria  DZ  DZA   12  ISO 3166-2:DZ
4  American Samoa  AS  ASM   16  ISO 3166-2:AS

In [31]:
datac_df.fillna(0, inplace=True)

In [32]:
df = datac_df.groupby(["Country", "A3C"])["Value"].agg(["median", "max", "min", "mean"]).reset_index()
df.columns = ["Country", "A3C", "Median", "Max", "Min", "Mean"]

In [33]:
def show_on_map(df, aggregate, title):
    fig = go.Figure(data=go.Choropleth(
        locations = df['A3C'],
        z = df[aggregate],
        text = df['Country'],
        colorscale = 'Rainbow',
        autocolorscale=False,
        reversescale=False,
        marker_line_color='darkgray',
        marker_line_width=0.5,
        colorbar_tickprefix = '%',
        colorbar_title = 'Violence %',
    ))

    fig.update_layout(
        title_text=title,
        geo=dict(
            showframe=False,
            showcoastlines=False,
            projection_type='equirectangular'
        ),
        annotations = [dict(
            x=0.55,
            y=0.1,
            xref='paper',
            yref='paper',
            text='Source: <a href="https://www.kaggle.com/andrewmvd/violence-against-women-and-girls">\
                Violence against Women and Girls</a>',
            showarrow = False
        )]
    )

    fig.show()

In [36]:
df.head()
df.to_csv("output_with_null.csv")

In [35]:
show_on_map(df, "Median", "Violence against Women and Girls - Country median level of violence[%]")

In [32]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="No education"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["median"]).reset_index()
df.columns = ["Country", "A3C", "Median"]
show_on_map(df, "Median", "Violence against Women with No education - Country Median level of violence [%]")

In [33]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="Higher"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["median"]).reset_index()
df.columns = ["Country", "A3C", "Median"]
show_on_map(df, "Median", "Violence against Women with Higher education - Country Median level of violence [%]")

In [34]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="Rural"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["median"]).reset_index()
df.columns = ["Country", "A3C", "Median"]
show_on_map(df, "Median", "Violence against Women in Rural Areas - Country Median level of violence [%]")

In [35]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="Urban"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["median"]).reset_index()
df.columns = ["Country", "A3C", "Median"]
show_on_map(df, "Median", "Violence against Women in Urban Areas - Country Median level of violence [%]")

In [36]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="15-24"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["median"]).reset_index()
df.columns = ["Country", "A3C", "Median"]
show_on_map(df, "Median", "Violence against Women age 15-24 - Country Median level of violence [%]")

In [37]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="25-34"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["median"]).reset_index()
df.columns = ["Country", "A3C", "Median"]
show_on_map(df, "Median", "Violence against Women age 25-34 - Country Median level of violence [%]")

In [38]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="35-49"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["median"]).reset_index()
df.columns = ["Country", "A3C", "Median"]
show_on_map(df, "Median", "Violence against Women age 35-49 - Country Median level of violence [%]")

# <a id='7'>Maximum violence percent per country</a>

In [39]:
df = datac_df.groupby(["Country", "A3C"])["Value"].agg(["median", "max", "min", "mean"]).reset_index()
df.columns = ["Country", "A3C", "Median", "Max", "Min", "Mean"]
show_on_map(df, "Max", "Violence against Women and Girls - Country maximum level of violence[%]")

In [40]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="No education"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["max"]).reset_index()
df.columns = ["Country", "A3C", "Max"]
show_on_map(df, "Max", "Violence against Women with No education - Country Max level of violence [%]")

In [41]:
filter_df = datac_df.loc[datac_df["Demographics Response"]=="Higher"]
df = filter_df.groupby(["Country", "A3C"])["Value"].agg(["max"]).reset_index()
df.columns = ["Country", "A3C", "Max"]
show_on_map(df, "Max", "Violence against Women with Higher education - Country Max level of violence [%]")

# <a id='8'>Summary of findings</a>


Some of the conclusions of the analysis are listed here:

* The most exposed age group is the 15-24, with a median of 17.50 and max of 80.1, while the absolute max is for 25-34 age group;

* Education level of girls and women is a very good predictor for violence against them: a higher education level will ensure a small median percent of 4.2%, while No education shows a 21.55% median and a 82% maximum violence percent.

* Employment factor also counts, from Unemployed to Employed for kind the median is varying from 14.55% to 20.15%. The women that work tend to be more exposed to violence.

* Marriage is slighly protecting if we look to the median, but married women are more exposed to higher level of violence. Widowed, divorced or separated have a smaller median than Married or living together group, as well as the max. But it might be also an age factor here.

* Women from Urban areas are less exposed to violence than women from Rural areas, both the median and the max values being smaller in Urban vs. Rural areas.

* Most exposed girls and women to extreme violence are from countries as Timor, Mali (subsaharian Africa), Chad, Eritreea, Afganistan and, surprisingly, Marroco. If the Marroco is a surprise, we should notice that the other countries are also either conflict zones or failes states, which is expected  to aggravate the life conditions of most exposed people.

* In the highest violence countries, the worst condition (i.e. exposure to higher levels of violence) is that of women or girls without education.





That's a lot better, now lets query results in order to make an exploratory data analysis acording to Age and Education.

In [42]:
fig = px.box(survey_df.query("demographics_question == 'Age'").sort_values('violence_any_reason',ascending=False),
            x      = 'country',
            y      = 'violence_any_reason',
            color  = 'gender',
            title  = '% of Respondents that agree with violence for any surveyed reason across Country and Gender',
            color_discrete_sequence = ['#4a00ba','#00ba82'],
            height = 650
        )

fig.update_xaxes(title='Country')
fig.update_yaxes(title='% Agrees: Violence is justified for any surveyed reason')
fig.show()

NameError: name 'px' is not defined